File: bindings/go/client/builders.go
package client

import pb "github.com/keystone-db/keystonedb/bindings/go/client/pb"

// PutRequestBuilder helps build PutRequest
type PutRequestBuilder struct {
	req *pb.PutRequest
}

// NewPutRequest creates a new PutRequestBuilder
func NewPutRequest(pk []byte) *PutRequestBuilder {
	return &PutRequestBuilder{
		req: &pb.PutRequest{
			PartitionKey:      pk,
			Item:              &pb.Item{Attributes: make(map[string]*pb.Value)},
			ExpressionValues:  make(map[string]*pb.Value),
		},
	}
}

// WithSortKey sets the sort key
func (b *PutRequestBuilder) WithSortKey(sk []byte) *PutRequestBuilder {
	b.req.SortKey = &sk
	return b
}

// WithAttribute adds an attribute to the item
func (b *PutRequestBuilder) WithAttribute(name string, value *pb.Value) *PutRequestBuilder {
	b.req.Item.Attributes[name] = value
	return b
}

// WithString adds a string attribute
func (b *PutRequestBuilder) WithString(name, value string) *PutRequestBuilder {
	b.req.Item.Attributes[name] = &pb.Value{
		Value: &pb.Value_StringValue{StringValue: value},
	}
	return b
}

// WithNumber adds a number attribute
func (b *PutRequestBuilder) WithNumber(name, value string) *PutRequestBuilder {
	b.req.Item.Attributes[name] = &pb.Value{
		Value: &pb.Value_NumberValue{NumberValue: value},
	}
	return b
}

// WithBool adds a boolean attribute
func (b *PutRequestBuilder) WithBool(name string, value bool) *PutRequestBuilder {
	b.req.Item.Attributes[name] = &pb.Value{
		Value: &pb.Value_BoolValue{BoolValue: value},
	}
	return b
}

// WithCondition sets a condition expression
func (b *PutRequestBuilder) WithCondition(condition string) *PutRequestBuilder {
	b.req.ConditionExpression = &condition
	return b
}

// Build returns the built PutRequest
func (b *PutRequestBuilder) Build() *pb.PutRequest {
	return b.req
}

// GetRequestBuilder helps build GetRequest
type GetRequestBuilder struct {
	req *pb.GetRequest
}

// NewGetRequest creates a new GetRequestBuilder
func NewGetRequest(pk []byte) *GetRequestBuilder {
	return &GetRequestBuilder{
		req: &pb.GetRequest{
			PartitionKey: pk,
		},
	}
}

// WithSortKey sets the sort key
func (b *GetRequestBuilder) WithSortKey(sk []byte) *GetRequestBuilder {
	b.req.SortKey = &sk
	return b
}

// Build returns the built GetRequest
func (b *GetRequestBuilder) Build() *pb.GetRequest {
	return b.req
}

// QueryRequestBuilder helps build QueryRequest
type QueryRequestBuilder struct {
	req *pb.QueryRequest
}

// NewQueryRequest creates a new QueryRequestBuilder
func NewQueryRequest(pk []byte) *QueryRequestBuilder {
	return &QueryRequestBuilder{
		req: &pb.QueryRequest{
			PartitionKey:     pk,
			ExpressionValues: make(map[string]*pb.Value),
		},
	}
}

// WithSortKeyEqual sets an equal condition on the sort key
func (b *QueryRequestBuilder) WithSortKeyEqual(value *pb.Value) *QueryRequestBuilder {
	b.req.SortKeyCondition = &pb.SortKeyCondition{
		Condition: &pb.SortKeyCondition_EqualTo{EqualTo: value},
	}
	return b
}

// WithSortKeyBeginsWith sets a begins_with condition on the sort key
func (b *QueryRequestBuilder) WithSortKeyBeginsWith(value *pb.Value) *QueryRequestBuilder {
	b.req.SortKeyCondition = &pb.SortKeyCondition{
		Condition: &pb.SortKeyCondition_BeginsWith{BeginsWith: value},
	}
	return b
}

// WithSortKeyBetween sets a between condition on the sort key
func (b *QueryRequestBuilder) WithSortKeyBetween(lower, upper *pb.Value) *QueryRequestBuilder {
	b.req.SortKeyCondition = &pb.SortKeyCondition{
		Condition: &pb.SortKeyCondition_Between{
			Between: &pb.BetweenCondition{Lower: lower, Upper: upper},
		},
	}
	return b
}

// WithLimit sets the query limit
func (b *QueryRequestBuilder) WithLimit(limit uint32) *QueryRequestBuilder {
	b.req.Limit = &limit
	return b
}

// WithIndex sets the index name
func (b *QueryRequestBuilder) WithIndex(indexName string) *QueryRequestBuilder {
	b.req.IndexName = &indexName
	return b
}

// WithScanForward sets the scan direction
func (b *QueryRequestBuilder) WithScanForward(forward bool) *QueryRequestBuilder {
	b.req.ScanForward = &forward
	return b
}

// Build returns the built QueryRequest
func (b *QueryRequestBuilder) Build() *pb.QueryRequest {
	return b.req
}

// ScanRequestBuilder helps build ScanRequest
type ScanRequestBuilder struct {
	req *pb.ScanRequest
}

// NewScanRequest creates a new ScanRequestBuilder
func NewScanRequest() *ScanRequestBuilder {
	return &ScanRequestBuilder{
		req: &pb.ScanRequest{
			ExpressionValues: make(map[string]*pb.Value),
		},
	}
}

// WithLimit sets the scan limit
func (b *ScanRequestBuilder) WithLimit(limit uint32) *ScanRequestBuilder {
	b.req.Limit = &limit
	return b
}

// WithSegment sets the scan segment
func (b *ScanRequestBuilder) WithSegment(segment, totalSegments uint32) *ScanRequestBuilder {
	b.req.Segment = &segment
	b.req.TotalSegments = &totalSegments
	return b
}

// Build returns the built ScanRequest
func (b *ScanRequestBuilder) Build() *pb.ScanRequest {
	return b.req
}

File: bindings/python/client/keystonedb/builders.py
"""Builder classes for KeystoneDB requests."""

from typing import Optional, Dict

from . import keystone_pb2 as pb


class PutRequestBuilder:
    """Builder for PutRequest.

    Example:
        >>> request = PutRequestBuilder(b"user#123") \\
        ...     .with_string("name", "Alice") \\
        ...     .with_number("age", "30") \\
        ...     .with_bool("active", True) \\
        ...     .build()
    """

    def __init__(self, partition_key: bytes):
        """Initialize builder with partition key.

        Args:
            partition_key: Partition key bytes
        """
        self.partition_key = partition_key
        self.sort_key: Optional[bytes] = None
        self.attributes: Dict[str, any] = {}
        self.condition: Optional[str] = None
        self.expression_values: Dict[str, any] = {}

    def with_sort_key(self, sort_key: bytes) -> "PutRequestBuilder":
        """Set sort key.

        Args:
            sort_key: Sort key bytes

        Returns:
            Self for chaining
        """
        self.sort_key = sort_key
        return self

    def with_string(self, name: str, value: str) -> "PutRequestBuilder":
        """Add string attribute.

        Args:
            name: Attribute name
            value: String value

        Returns:
            Self for chaining
        """
        self.attributes[name] = {"string_value": value}
        return self

    def with_number(self, name: str, value: str) -> "PutRequestBuilder":
        """Add number attribute.

        Args:
            name: Attribute name
            value: Number as string (for precision)

        Returns:
            Self for chaining
        """
        self.attributes[name] = {"number_value": value}
        return self

    def with_bool(self, name: str, value: bool) -> "PutRequestBuilder":
        """Add boolean attribute.

        Args:
            name: Attribute name
            value: Boolean value

        Returns:
            Self for chaining
        """
        self.attributes[name] = {"bool_value": value}
        return self

    def with_binary(self, name: str, value: bytes) -> "PutRequestBuilder":
        """Add binary attribute.

        Args:
            name: Attribute name
            value: Binary value

        Returns:
            Self for chaining
        """
        self.attributes[name] = {"binary_value": value}
        return self

    def with_condition(self, condition: str) -> "PutRequestBuilder":
        """Set condition expression.

        Args:
            condition: Condition expression string

        Returns:
            Self for chaining
        """
        self.condition = condition
        return self

    def build(self) -> pb.PutRequest:
        """Build the PutRequest.

        Returns:
            PutRequest message
        """
        item = pb.Item(attributes={
            name: pb.Value(**val) for name, val in self.attributes.items()
        })

        request = pb.PutRequest(
            partition_key=self.partition_key,
            item=item,
        )

        if self.sort_key is not None:
            request.sort_key = self.sort_key
        if self.condition is not None:
            request.condition_expression = self.condition
        if self.expression_values:
            for k, v in self.expression_values.items():
                request.expression_values[k].CopyFrom(pb.Value(**v))

        return request


class GetRequestBuilder:
    """Builder for GetRequest.

    Example:
        >>> request = GetRequestBuilder(b"user#123").build()
        >>> # With sort key:
        >>> request = GetRequestBuilder(b"user#123") \\
        ...     .with_sort_key(b"profile") \\
        ...     .build()
    """

    def __init__(self, partition_key: bytes):
        """Initialize builder with partition key.

        Args:
            partition_key: Partition key bytes
        """
        self.partition_key = partition_key
        self.sort_key: Optional[bytes] = None

    def with_sort_key(self, sort_key: bytes) -> "GetRequestBuilder":
        """Set sort key.

        Args:
            sort_key: Sort key bytes

        Returns:
            Self for chaining
        """
        self.sort_key = sort_key
        return self

    def build(self) -> pb.GetRequest:
        """Build the GetRequest.

        Returns:
            GetRequest message
        """
        request = pb.GetRequest(partition_key=self.partition_key)
        if self.sort_key is not None:
            request.sort_key = self.sort_key
        return request


class QueryRequestBuilder:
    """Builder for QueryRequest.

    Example:
        >>> request = QueryRequestBuilder(b"org#acme") \\
        ...     .with_limit(10) \\
        ...     .with_index("status-index") \\
        ...     .build()
    """

    def __init__(self, partition_key: bytes):
        """Initialize builder with partition key.

        Args:
            partition_key: Partition key bytes
        """
        self.partition_key = partition_key
        self.sort_key_condition: Optional[Dict] = None
        self.filter_expression: Optional[str] = None
        self.expression_values: Dict[str, any] = {}
        self.index_name: Optional[str] = None
        self.limit: Optional[int] = None
        self.exclusive_start_key: Optional[Dict] = None
        self.scan_forward: Optional[bool] = None

    def with_sort_key_equal(self, value) -> "QueryRequestBuilder":
        """Set sort key equal condition.

        Args:
            value: Value to match

        Returns:
            Self for chaining
        """
        self.sort_key_condition = {"equal_to": value}
        return self

    def with_sort_key_begins_with(self, value) -> "QueryRequestBuilder":
        """Set sort key begins_with condition.

        Args:
            value: Prefix to match

        Returns:
            Self for chaining
        """
        self.sort_key_condition = {"begins_with": value}
        return self

    def with_sort_key_between(self, lower, upper) -> "QueryRequestBuilder":
        """Set sort key between condition.

        Args:
            lower: Lower bound
            upper: Upper bound

        Returns:
            Self for chaining
        """
        self.sort_key_condition = {"between": {"lower": lower, "upper": upper}}
        return self

    def with_limit(self, limit: int) -> "QueryRequestBuilder":
        """Set query limit.

        Args:
            limit: Maximum items to return

        Returns:
            Self for chaining
        """
        self.limit = limit
        return self

    def with_index(self, index_name: str) -> "QueryRequestBuilder":
        """Set index name.

        Args:
            index_name: Name of index to query

        Returns:
            Self for chaining
        """
        self.index_name = index_name
        return self

    def with_scan_forward(self, forward: bool) -> "QueryRequestBuilder":
        """Set scan direction.

        Args:
            forward: True for ascending, False for descending

        Returns:
            Self for chaining
        """
        self.scan_forward = forward
        return self

    def build(self):
        """Build the QueryRequest.

        Returns:
            QueryRequest message (once protobuf is generated)
        """
        return {
            "partition_key": self.partition_key,
            "sort_key_condition": self.sort_key_condition,
            "filter_expression": self.filter_expression,
            "expression_values": self.expression_values,
            "index_name": self.index_name,
            "limit": self.limit,
            "exclusive_start_key": self.exclusive_start_key,
            "scan_forward": self.scan_forward,
        }


class ScanRequestBuilder:
    """Builder for ScanRequest.

    Example:
        >>> request = ScanRequestBuilder() \\
        ...     .with_limit(100) \\
        ...     .with_segment(0, 4) \\
        ...     .build()
    """

    def __init__(self):
        """Initialize empty builder."""
        self.filter_expression: Optional[str] = None
        self.expression_values: Dict[str, any] = {}
        self.limit: Optional[int] = None
        self.exclusive_start_key: Optional[Dict] = None
        self.index_name: Optional[str] = None
        self.segment: Optional[int] = None
        self.total_segments: Optional[int] = None

    def with_limit(self, limit: int) -> "ScanRequestBuilder":
        """Set scan limit.

        Args:
            limit: Maximum items to return

        Returns:
            Self for chaining
        """
        self.limit = limit
        return self

    def with_segment(self, segment: int, total_segments: int) -> "ScanRequestBuilder":
        """Set parallel scan segment.

        Args:
            segment: Segment number (0-based)
            total_segments: Total number of parallel segments

        Returns:
            Self for chaining
        """
        self.segment = segment
        self.total_segments = total_segments
        return self

    def build(self):
        """Build the ScanRequest.

        Returns:
            ScanRequest message (once protobuf is generated)
        """
        return {
            "filter_expression": self.filter_expression,
            "expression_values": self.expression_values,
            "limit": self.limit,
            "exclusive_start_key": self.exclusive_start_key,
            "index_name": self.index_name,
            "segment": self.segment,
            "total_segments": self.total_segments,
        }
